akka-stream/src/main/scala/akka/stream/scaladsl/SourceRef.scala

```diff
@@ -0,0 +1,272 @@
+
+          tryPush(payload)
+          triggerCumulativeDemand()
+
+        case (sender, StreamRefs.RemoteStreamCompleted(seqNr)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in RemoteSinkCompleted")
+          observeAndValidateSequenceNr(seqNr, "Illegal sequence nr in RemoteSinkCompleted")
+          log.debug("[{}] The remote stream has completed, completing as well...", stageActorName)
+
+          self.unwatch(sender)
+          completeStage()
+
+        case (sender, StreamRefs.RemoteStreamFailure(reason)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in RemoteSinkFailure")
+          log.warning("[{}] The remote stream has failed, failing (reason: {})", stageActorName, reason)
+
+          self.unwatch(sender)
+          failStage(StreamRefs.RemoteStreamRefActorTerminatedException(s"Remote stream (${sender.path}) failed, reason: $reason"))
+
+        case (_, Terminated(ref)) ⇒
+          if (getPartnerRef == ref)
+            failStage(StreamRefs.RemoteStreamRefActorTerminatedException(s"The remote partner ${getPartnerRef} has terminated! " +
+              s"Tearing down this side of the stream as well."))
+        // else this should not have happened, and we ignore this -- someone may have been doing weird things!
+      }
+
+      def tryPush(): Unit =
+        if (isAvailable(out) && receiveBuffer.nonEmpty) {
+          val elem = receiveBuffer.dequeue()
+          push(out, elem)
+        } else canPush = true
+
+      def tryPush(payload: Any): Unit = {
+        localRemainingRequested -= 1
+        if (canPush) {
+          canPush = false
+
+          if (receiveBuffer.nonEmpty) {
+            val elem = receiveBuffer.dequeue()
+            push(out, elem)
+            receiveBuffer.enqueue(payload.asInstanceOf[T])
+          } else {
+            push(out, payload.asInstanceOf[T])
+          }
+        } else {
+          if (receiveBuffer.isFull) throw new IllegalStateException(s"Attempted to overflow buffer! Capacity: ${receiveBuffer.capacity}, incoming element: $payload, localRemainingRequested: ${localRemainingRequested}, localCumulativeDemand: ${localCumulativeDemand}")
+          receiveBuffer.enqueue(payload.asInstanceOf[T])
+        }
+      }
+
+      @throws[StreamRefs.InvalidPartnerActorException]
+      def observeAndValidateSender(partner: ActorRef, msg: String): Unit =
+        if (partnerRef.isEmpty) {
+          log.debug("Received first message from {}, assuming it to be the remote partner for this stage", partner)
+          partnerRef = OptionVal(partner)
+          self.watch(partner)
+        } else if (partnerRef.isDefined && partner != getPartnerRef) {
+          val ex = StreamRefs.InvalidPartnerActorException(partner, getPartnerRef, msg)
+          partner ! StreamRefs.RemoteStreamFailure(ex.getMessage)
+          throw ex
+        }
+
+      @throws[StreamRefs.InvalidSequenceNumberException]
```

Comment:
Nitpick: Really needed? ;)

Reply:
lol yeah, not really ;-) Will remove, not really our style